scipy/sparse/csgraph/_disjoint_set.py

```diff
@@ -0,0 +1,138 @@
+"""
+Disjoint set data structure
+"""
+
+
+import numpy as np
+
+
+class DisjointSet:
+    """ Disjoint set data structure for incremental connectivity queries.
+
+    .. versionadded: 1.6.0
+
+    Attributes
+    ----------
+    n_nodes : int
+        The number of nodes in the set.
+    n_components : int
+        The number of components/subsets.
+
+    Methods
+    -------
+    union
+    find
+
+    Notes
+    -----
+    This class implements the disjoint set [1]_, also known as the *union-find*
+    data structure. The *find* method implements the *path compression*
+    variant. The *union* method implements the *union by size* variant.
+
+    References
+    ----------
+    .. [1] https://en.wikipedia.org/wiki/Disjoint-set_data_structure
+
+    See Also
+    --------
+    connected_components : Analyze the connected components of a sparse graph.
+
+    Examples
+    --------
+    >>> from scipy.sparse.csgraph import DisjointSet
+
+    Initialize a disjoint set:
+
+    >>> dis = DisjointSet()
+
+    Merge some subsets:
+
+    >>> dis.union(0, 1)
+    True
+    >>> dis.union(2, 3)
+    True
+    >>> dis.union(3, 3)
+    False
+
+    Find a root node:
+
+    >>> dis.find(1)
+    0
+
+    """
+    def __init__(self):
+        self.n_nodes = 0
+        self.n_components = 0
+        self._sizes = {}
+        self._parents = {}
+
+    def find(self, x):
+        """Find the root node of `x`.
+
+        Parameters
+        ----------
+        x : int
+            Input node.
+
+        Returns
+        -------
+        root : int
+            Root node of `x`.
+        """
+        if not np.issubdtype(type(x), np.integer):
+            raise TypeError("`x` must be an integer")
+
+        if x not in self._parents:
+            self._sizes[x] = 1
+            self._parents[x] = x
+            self.n_nodes += 1
+            self.n_components += 1
+            return x
+
+        parents = self._parents
+        parent = parents[x]
+        while parent != parents[parent]:
+            parent = parents[parent]
+        parents[x] = parent
```

Comment:
You mention in the documentation that you implement path compression, but from a quick look, that doesn't quite seem to be the case (at least not in the sense I'm familiar with); while the parent is indeed updated for the input node `x`, it's not updated for anything else along the way (which would require a recursion, explicitly keeping track of the path, two iterations, or something along those lines).

Reply:
Another great catch. I have now implemented "path-halving" and updated the docs.